### ISSUE TITLE
Support internal links

### ISSUE THREAD

**Issue by frankgerhardt** [nextcloud/notes#386]
This is somewhat related to #158 and #173 because it's about Markdown and links, but specifically about **internal** links, i.e. links from one note to another note. 

E.g. I have two notes called `First` and `Second` and then I have a link in the first note like this `[link to Second](Second)`. Then that link should be clickable and go to the second note.

**Comment by stefan-niedermann** [comment]
I think there was a similar proposeal some years ago for the android client: https://github.com/stefan-niedermann/nextcloud-notes/pull/202

**Comment by stefan-niedermann** [comment]
On the android client we went for the IDs. That's simply because the title is a often changed element which would invalidate all previously linked notes **and** would cause conflicts (Often titles are the same like 'todo' or so on multiple notes).

Just copy the number from the url of the target note and use this as link target.

Example:

![grafik](https://user-images.githubusercontent.com/4741199/73884649-a7f5f680-4866-11ea-975b-9e7d6a416ae4.png)

To make a link to this note, just use `[note from screenshot](46106)`

I just tested this behavior and it works in the preview mode of the notes web app, too.

**Comment by stefan-niedermann** [comment]
Therefore i think this issue is basically already solved, **but** it might be a separate issue to suggest notes while typing in brackets of a link and insert the ID automatically on click (similar to how GitHub @-mentions are working).

**Comment by korelstar** [comment]
Perfect. I'm closing this for now. Further feature request should go in a new issue.

**Comment by frankgerhardt** [comment]
Yes, it works. But on the mobile app there is no way to find the id like 46106. That makes it rather unfriendly to use. Well, in other words, it's impossible to create a link without the id. I added a feature request https://github.com/stefan-niedermann/nextcloud-notes/issues/900

**Comment by frankgerhardt** [comment]
This my issue is now closed but I'm coming back to say that for me it is unhappily closed. 

I have used Notes in three ways, in the web UI, in the Notes Android app and in the file system using a local editor with markdown syntax highlighting. The last use case is my most common one and it sucks the most when it comes to links. Simply because the note id has to be used which I don't know in the offline file system use case. 

The linking using ids is a leaky abstraction. Text and page names are the content layer and ids on the implementation layer. Mixing both layers in the links is a design mistake. I wish you redesign this some day. 

The workaround is to create dummy notes called NewNote1, NewNote2 etc., add as content "my id is xxxx" and use those for new Notes when working in the file system. I hope this ugly shameful example serves as a motivation for fixing the design.

**Comment by stefan-niedermann** [comment]
> fixing the design

Mayve you can propose a solution for the following issue which lead us to use IDs in favor of Note titles for link tagets:

What happens if the note title changes?
 - One could say "sure, update all notes which have a link to it" 
  This will only work for links which are in Nextcloud but not for links which are e. g. in a PDF document or anywhere else. Also it would not work in your offline local editor scenario.
- One could also maintain a history of a previohs file names and redirect the links - but this would lead to conflicts as soon asanother note changes the title to something thar was already used before

**Comment by frankgerhardt** [comment]
Don't try to fix anything when a link could possibly break. Broken links are a fact on the Internet. You can't fix the way the world doesn't work. Users know what they are doing. A deluxe version could update links in the same directory at most. Russian proverb: good intentions pave the way to hell. 

```
# The header of a note
123456

Above is the ID in case I want to make a link to it.

Lore ipsum...
```

For offline editing I now put the note id in the second line of every note. How bad is that!?

**Comment by stefan-niedermann** [comment]
> Broken links are a fact on the Internet.

Well, links in the internet only break if one changes the URI. Just by editing the title of a website, no link will break :sl

**Comment by caseyhoward** [comment]
Editing the title of a markdown note won't beat links either. It's only when you edit the filename itself, just like websites.

**Comment by stefan-niedermann** [comment]
>  when you edit the filename itself

Note title == filename. When you change the title of a note, the file will also get renamed.

**Comment by caseyhoward** [comment]
I forgot about that. I stopped using Notes a while ago because of the linking situation. I just use markdown files but I use VSCode to browse them instead of Nextcloud because of this issue.

Couldn't a solution be to not sync the file name with the title?

**Comment by stefan-niedermann** [comment]
This would need the title
- either to be stored as some kind of meta data (which we can not do because we want to keep it compatible with DAV-based file syncing like the default Nextcloud Desktop client you are using)
- or to always use the first line of the note as title what would lead to odd behavior because we suddenly would have kind of two titles

**Comment by mwerle** [comment]
Just my 2c, as unfortunately the Tomdroid project appears to be dead. 

For me the killer feature of Tomdroid was always the auto linking of notes. simply by adding text to a note which matched the title of another note was sufficient to create a link. 

Change a notes title, link breaks. Shoganai, user problem. 

I would _love_ for that to be added to Notes. Can always have multiple link types - explicit ones with the note ID (permalink) and the implicit ones i just described. 

killer for me is NOT having to manually write complicated markdown link syntax on my phone, which is where i mostly use this.
